src/to_wiki.js

```diff
@@ -74,18 +76,66 @@ class WikiTextSerializerState {
    }
 
    // Apply marks to inline text.
+   // Find lengths of each continuous mark use and store in Map. (This can't be
+   // done simultaneously with the next step, because we need to know the full
+   // length of each mark in order to put them on in the right order.)
+   // For each text node (this is done in the renderer for text nodes):
+   //   1. Find closing marks
+   //   2. Pop marks off the stack until all the closing marks are gone. Close
+   //   all marks popped off.
+   //   3. Generate start characters for all marks present on this node and
+   //   not already on stack. Push onto the stack in the order generated. This
+   //   will include any marks popped in step 2 which are supposed to be on the
+   //   node.
+   //
    inline(node) {
+      const markLengths = new Map()
+      // Compute mark lengths
+      let findMarkLengths = (node) => {
+         // Gather all the marks on this node (by name, so they will match the
+         // marks from previous nodes).
+         let marks = node.marks.map(m => m.type.name)
+
+         // Marks that exist for the current node, but not in markLengths are
+         // new marks, so they should be added and set to 0. We'll add the
+         // length of the current text to both the new and old marks next.
+         marks.forEach(m => {
+            if (!markLengths.has(m)) {
+               markLengths.set(m, 0)
+            }
+         });
+         markLengths.forEach((v, k) => markLengths.set(k, v + node.text.length))
+
+         // Marks in markLengths that do not exist on the current node are no
+         // longer needed. Remove them. The ones that do exist, however, should
+         // be in the list of lengths for the marks on this node.
+         let lengths = new Map()
+         markLengths.forEach((v, k) => {
+            if (marks.indexOf(k) < 0) {
+               markLengths.delete(k)
+            } else {
+               lengths.set(k, v)
+            }
+         });
+         // Store the list of lengths for later use by the renderer.
+         this.lengths.set(node, lengths)
+      };
+
+      const children = []
+      node.forEach(c => children.push(c))
+      children.reverse()
+      children.forEach(findMarkLengths)
```

Comment:
[curious] Does this need to run `findMarkLengths` recursively on children's children? Or is there enough information on the direct children to figure out what you need to know?